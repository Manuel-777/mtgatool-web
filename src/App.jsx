/* eslint-disable react/prop-types */
import React from "react";
import { BrowserRouter as Router, Route, Switch, Link } from "react-router-dom";
import { hot } from "react-hot-loader/root";
import db from "./shared/database";

import Video from "./components/video";
import {
  WrapperInner,
  WrapperOuter,
  WrapperOuterLight
} from "./components/wrapper";
import TopNav from "./components/topnav";
import Footer from "./components/footer";
import NotFound from "./components/notfound";
import ReleaseNotes from "./components/release-notes";
import Metagame from "./components/metagame";
import Register from "./components/register";
import CardHover from "./components/card-hover";
import {WebProvider} from "./web-provider";

// Import once so all CSS can use it thanks to webpack magic
// eslint-disable-next-line no-unused-vars
import { STATE_IDLE, STATE_DOWNLOAD, STATE_ERROR } from "./constants";
import sharedcss from "./shared.css";
import css from "./app.css";
import downloadButton from "./images/badge_download.png";
import keyArt from "./images/key-art.jpg";

const DESCRIPTION_TEXT = `MTG Arena Tool is a collection browser, a deck tracker and a statistics manager. Explore which decks you played against and what other players are brewing. MTG Arena Tool is all about improving your Magic Arena experience.`;

const FEATURE_A_TITLE = `Powerful stats`;
const FEATURE_A_TEXT = `With MTGA Tool you will be able to browse your match history and see which cards and colors your opponents played. This is extremely helpful to understand why your deck is struggling or winning against other opponents. See how much progress you have made with your collection, browse your new cards acquired and explore how many packs and wildcards you need to finish each of your deck builds.`;
const FEATURE_B_TITLE = `Explore and learn`;
const FEATURE_B_TEXT = `Browse what others are playing and how they perform, be it on constructed events, drafts and ranked. You would be able to filter decks by colors, events or results. Take your game to the next level and brag your winning deck with everyone else!`;
const FEATURE_C_TITLE = `Not only a Deck Tracker`;
const FEATURE_C_TEXT = `Even if you dislike deck trackers, you can find MTGA Tool useful to keep record of your matches, browse through your collection or even check other people's decks. Just disable the deck tracker overlay to keep it running only in background, you wonâ€™t miss a thing anyway.`;

const RELEASE_DESC = "v2.14.2, Released November 29th, 2019";

const DATABASE_URL = "https://mtgatool.com/database/";

function App() {
  const artist = "Bedevil by Seb Mckinnon";
  const [image, setImage] = React.useState(keyArt);
  const [queryState, setQueryState] = React.useState(0);

  const getDatabase = () => {
    if (localStorage.databaseTime) {
      const dbJson = JSON.parse(localStorage.database);
      console.log("database from cache: v" + dbJson.version);
      db.setDatabase(localStorage.database);
    }
    if (
      !localStorage.database ||
      new Date(Date.now() - 864e5) < new Date(localStorage.databaseTime)
    ) {
      setQueryState(STATE_DOWNLOAD);
      const xhr = new XMLHttpRequest();
      xhr.onload = () => {
        if (xhr.status !== 200) {
          setQueryState(xhr.status);
        } else {
          try {
            console.log("Database download ok!");
            localStorage.database = xhr.responseText;
            localStorage.databaseTime = new Date();
            db.setDatabase(xhr.responseText);
            // Query state could be in redux
            setQueryState(STATE_IDLE);
          } catch (e) {
            console.log(e);
            setQueryState(STATE_ERROR);
          }
        }
      };
      xhr.onreadystatechange = () => {
        if (xhr.readyState === 4 && queryState !== STATE_ERROR) {
          setQueryState(STATE_IDLE);
        }
      };
      xhr.open("GET", DATABASE_URL);
      xhr.send();
    }
  };

  React.useEffect(() => {
    getDatabase();
  }, []);

  const wrapperStyle = {
    backgroundImage: `url("${image}")`
  };

  return (
    <WebProvider>
      <Router>
        <div style={wrapperStyle} className={css["wrapper-image"]} />
        <TopNav artist={artist} />
        <CardHover />
        <Switch>
          <Route exact path="/">
            <Home setImage={setImage} />
          </Route>
          <Route path="/metagame">
            <Metagame setImage={setImage} />
          </Route>
          <Route exact path="/register">
            <Register setImage={setImage} />
          </Route>
          <Route exact path="/release-notes">
            <ReleaseNotes setImage={setImage} />
          </Route>
          <Route>
            <NotFound setImage={setImage} />
          </Route>
        </Switch>
        <Footer />
      </Router>
    </WebProvider>
  );
}

function Home(props) {
  const { setImage } = props;
  React.useEffect(() => {
    setImage(keyArt);
  }, []);

  return (
    <>
      <WrapperOuter>
        <WrapperInner>
          <div className={css["home-desc"]}>
            <div
              className={`${sharedcss["text-description"]} ${sharedcss["text-light"]}`}
            >
              {DESCRIPTION_TEXT}
            </div>
            <a
              className={css["download-button"]}
              href="https://github.com/Manuel-777/MTG-Arena-Tool/releases/latest"
            >
              <img
                className={css["download-img"]}
                alt=""
                src={downloadButton}
              />
            </a>
            <div className={css["home-desc-small"]}>{RELEASE_DESC}</div>
            <Link
              to="/release-notes"
              style={{ marginBottom: "10px" }}
              className={css["nav-logo"]}
            >
              Release Notes
            </Link>
            <div className={css["home-desc-small"]}>
              <i>* No account required</i>
            </div>
          </div>
        </WrapperInner>
      </WrapperOuter>
      <WrapperOuter style={{ paddingBottom: "64px" }}>
        <Video />
      </WrapperOuter>
      <WrapperOuterLight>
        <WrapperInner>
          <div className={sharedcss["text-title"]}>{FEATURE_A_TITLE}</div>
          <div className={sharedcss["text-description"]}>{FEATURE_A_TEXT}</div>
        </WrapperInner>
      </WrapperOuterLight>
      <WrapperOuterLight>
        <WrapperInner>
          <div className={sharedcss["text-title"]}>{FEATURE_B_TITLE}</div>
          <div className={sharedcss["text-description"]}>{FEATURE_B_TEXT}</div>
        </WrapperInner>
      </WrapperOuterLight>
      <WrapperOuterLight>
        <WrapperInner>
          <div className={sharedcss["text-title"]}>{FEATURE_C_TITLE}</div>
          <div className={sharedcss["text-description"]}>{FEATURE_C_TEXT}</div>
        </WrapperInner>
      </WrapperOuterLight>
    </>
  );
}

export default hot(App);
