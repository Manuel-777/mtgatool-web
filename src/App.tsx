/* eslint-disable react/prop-types */
import { useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import { ToolDbClient } from "tool-db";

import { AppState } from "./redux/stores/webStore";

import TopNav from "./components/topnav";
import Footer from "./components/footer";
import Home from "./components/home";
import NotFound from "./components/notfound";
import ReleaseNotes from "./components/release-notes";
import Register from "./components/register";
import Docs from "./components/docs";
import Database from "./components/database";
import Loading from "./components/loading";
import CookiesSign from "./components/cookies";
import { WrapperOuter } from "./components/wrapper";

import { DB_SERVER } from "./constants";

import keyArt from "./assets/images/key-art-new.jpg";
import notFoundArt from "./assets/images/404.jpg";
import "./App.css";

declare global {
  interface Window {
    toolDb: ToolDbClient;
    toolDbInitialized: boolean;
  }
}

export default function App(): JSX.Element {
  const [artData, setArtData] = useState("");
  const { databaseVersion, backImage } = useSelector(
    (state: AppState) => state.web
  );
  const [imageUrl, setImageUrl] = useState(backImage);

  useEffect(() => {
    if (!window.toolDbInitialized) {
      window.toolDb = new ToolDbClient(DB_SERVER);
      window.toolDbInitialized = true;
    }
  }, []);

  useEffect(() => {
    if (backImage === keyArt || backImage === "") {
      setImageUrl(keyArt);
      setArtData("Sublime Epiphany by Lindsey Look");
    } else if (backImage === notFoundArt) {
      setImageUrl(notFoundArt);
      setArtData("Totally Lost by David Palumbo");
    } else if (backImage && typeof backImage !== "string") {
      setImageUrl(backImage.images.art_crop);
      setArtData(`${backImage.name} by ${backImage.artist}`);
    }
  }, [backImage]);

  const wrapperStyle = {
    backgroundImage: `url("${imageUrl}")`,
  };

  return (
    <>
      <Database />
      <Loading />
      <CookiesSign />
      <Router>
        <div style={wrapperStyle} className="wrapper-image" />
        <TopNav artist={artData} />

        {databaseVersion !== 0 ? (
          <Switch>
            <Route exact path="/">
              <Home />
            </Route>
            <Route exact path="/register">
              <Register />
            </Route>
            <Route exact path="/release-notes">
              <ReleaseNotes />
            </Route>
            <Route path="/docs">
              <Docs />
            </Route>
            <Route>
              <NotFound />
            </Route>
          </Switch>
        ) : (
          <WrapperOuter>
            <div className="loading">Loading..</div>
          </WrapperOuter>
        )}
        <Footer />
      </Router>
    </>
  );
}
