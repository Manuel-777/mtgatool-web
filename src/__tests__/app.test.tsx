/* eslint-env jest */
import React from "react";
import App from "../App";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import store from "../redux/stores/webStore";

test("Renders", () => {
  const component = renderer.create(
    <Provider store={store}>
      <App />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  /*
  // manually trigger the callback
  tree.props.onMouseEnter();
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  tree.props.onMouseLeave();
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  */
});
