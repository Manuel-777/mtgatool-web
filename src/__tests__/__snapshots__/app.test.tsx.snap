// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
Array [
  <div
    className="loadingDiv"
    style={
      Object {
        "animation": "none",
        "display": "none",
      }
    }
  />,
  <div
    className="cookieBanner"
  >
    <div
      className="cookieText"
    >
      This website uses cookies to improve use experience by storing data on your browser and uses analytics data of Google Analytics services. By using this website you agree with our Privacy policy and terms of service.
    </div>
    <div
      className="buttonSimple"
      onClick={[Function]}
      style={Object {}}
    >
      I agree
    </div>
  </div>,
  <div
    className="wrapperImage"
    style={
      Object {
        "backgroundImage": "url(\\"test-file-stub\\")",
      }
    }
  />,
  <div
    className="topNavContainer"
  >
    <div
      className="topNav"
    >
      <div
        className="navLogoContainer"
      >
        <a
          className="navLogo"
          href="/"
          onClick={[Function]}
        />
      </div>
      <div
        className="navArtist"
      >
        Sublime Epiphany by Lindsey Look
      </div>
      <div
        className="navDivider"
      />
      <div
        className="navLinks"
      >
        <a
          className="navLinkA"
          href="/metagame"
          onClick={[Function]}
        >
          Metagame
        </a>
        <a
          className="navLinkA"
          href="/register"
          onClick={[Function]}
        >
          Register
        </a>
        <a
          className="navLinkA"
          href="/release-notes"
          onClick={[Function]}
        >
          Release Notes
        </a>
        <a
          className="navLinkA"
          href="/docs"
          onClick={[Function]}
        >
          Docs
        </a>
      </div>
    </div>
  </div>,
  <div
    className="card-hover-dfc"
    style={
      Object {
        "backgroundImage": "url(test-file-stub)",
        "opacity": 0,
      }
    }
  />,
  <div
    className="card-hover-main"
    style={
      Object {
        "backgroundImage": "url(https://img.scryfall.com/cardsundefined",
        "opacity": 0,
      }
    }
  />,
  <div
    className="wrapperOuter"
  >
    <div
      className="loading"
    >
      Loading..
    </div>
  </div>,
  <div
    className="footerMain"
  >
    <div
      className="footerWrapper"
    >
      <div
        className="footerColumn"
      >
        <img
          alt="MTG Arena Tool"
          className="footerLogo"
          src="test-file-stub"
          style={
            Object {
              "margin": "auto ",
            }
          }
        />
      </div>
      <div
        className="footerColumn"
      >
        <div
          className="footerColumnTitle"
        >
          MTG Arena Tool
        </div>
        <a
          href="mailto:mtgatool@gmail.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          Contact
        </a>
        <a
          href="https://mtgatool.com/database/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Database
        </a>
        <a
          href="https://mtgatool.com/docs/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Documentation
        </a>
        <a
          href="https://github.com/Manuel-777/MTG-Arena-Tool/releases"
          rel="noopener noreferrer"
          target="_blank"
        >
          Releases
        </a>
      </div>
      <div
        className="footerColumn"
      >
        <div
          className="footerColumnTitle"
        >
          Support us
        </div>
        <a
          href="https://www.paypal.me/ManuelEtchegaray/10"
          rel="noopener noreferrer"
          target="_blank"
        >
          Donate
        </a>
        <a
          href="https://www.patreon.com/mtgatool"
          rel="noopener noreferrer"
          target="_blank"
        >
          Patreon
        </a>
      </div>
      <div
        className="footerColumn"
      >
        <div
          className="footerColumnTitle"
        >
          Developers
        </div>
        <a
          href="https://github.com/Manuel-777/MTG-Arena-Tool"
          rel="noopener noreferrer"
          target="_blank"
        >
          App Github
        </a>
        <a
          href="https://github.com/mtgatool/mtgatool-web/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Website Github
        </a>
      </div>
      <div
        className="footerColumn"
      >
        <div
          className="footerColumnTitle"
        >
          Places
        </div>
        <a
          href="https://discord.gg/K9bPkJy"
          rel="noopener noreferrer"
          target="_blank"
        >
          Discord
        </a>
        <a
          href="https://twitter.com/MEtchegaray7"
          rel="noopener noreferrer"
          target="_blank"
        >
          Twitter
        </a>
        <a
          href="https://magic.wizards.com/mtgarena"
          rel="noopener noreferrer"
          target="_blank"
        >
          MTG: Arena
        </a>
      </div>
    </div>
    <div
      className="footerWrapper"
    >
      <div
        className="footerText"
      >
        Portions of MTG Arena Tool are unofficial Fan Content permitted under the Fan Content Policy.
        <br />
        The literal and graphical information presented on this site about Magic: The Gathering, including card images, the mana symbols, and Oracle text, is copyright Wizards of the Coast, LLC, a subsidiary of Hasbro, Inc. MTG Arena Tool is not produced by, endorsed by, supported by, or affiliated with Wizards of the Coast.
        <br />
        ©Wizards of the Coast LLC.
        <br />
        All other content ©2020 MTG Arena Tool.
      </div>
    </div>
  </div>,
]
`;

exports[`Renders 1`] = `
Array [
  <div
    className="loadingDiv"
    style={
      Object {
        "animation": "none",
        "display": "none",
      }
    }
  />,
  <div
    className="cookieBanner"
  >
    <div
      className="cookieText"
    >
      This website uses cookies to improve use experience by storing data on your browser and uses analytics data of Google Analytics services. By using this website you agree with our Privacy policy and terms of service.
    </div>
    <div
      className="buttonSimple"
      onClick={[Function]}
      style={Object {}}
    >
      I agree
    </div>
  </div>,
  <div
    className="wrapperImage"
    style={
      Object {
        "backgroundImage": "url(\\"test-file-stub\\")",
      }
    }
  />,
  <div
    className="topNavContainer"
  >
    <div
      className="topNav"
    >
      <div
        className="navLogoContainer"
      >
        <a
          className="navLogo"
          href="/"
          onClick={[Function]}
        />
      </div>
      <div
        className="navArtist"
      >
        Sublime Epiphany by Lindsey Look
      </div>
      <div
        className="navDivider"
      />
      <div
        className="navLinks"
      >
        <a
          className="navLinkA"
          href="/metagame"
          onClick={[Function]}
        >
          Metagame
        </a>
        <a
          className="navLinkA"
          href="/register"
          onClick={[Function]}
        >
          Register
        </a>
        <a
          className="navLinkA"
          href="/release-notes"
          onClick={[Function]}
        >
          Release Notes
        </a>
        <a
          className="navLinkA"
          href="/docs"
          onClick={[Function]}
        >
          Docs
        </a>
      </div>
    </div>
  </div>,
  <div
    className="card-hover-dfc"
    style={
      Object {
        "backgroundImage": "url(test-file-stub)",
        "opacity": 0,
      }
    }
  />,
  <div
    className="card-hover-main"
    style={
      Object {
        "backgroundImage": "url(https://img.scryfall.com/cardsundefined",
        "opacity": 0,
      }
    }
  />,
  <div
    className="wrapperOuter"
  >
    <div
      className="loading"
    >
      Loading..
    </div>
  </div>,
  <div
    className="footerMain"
  >
    <div
      className="footerWrapper"
    >
      <div
        className="footerColumn"
      >
        <img
          alt="MTG Arena Tool"
          className="footerLogo"
          src="test-file-stub"
          style={
            Object {
              "margin": "auto ",
            }
          }
        />
      </div>
      <div
        className="footerColumn"
      >
        <div
          className="footerColumnTitle"
        >
          MTG Arena Tool
        </div>
        <a
          href="mailto:mtgatool@gmail.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          Contact
        </a>
        <a
          href="https://mtgatool.com/database/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Database
        </a>
        <a
          href="https://mtgatool.com/docs/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Documentation
        </a>
        <a
          href="https://github.com/Manuel-777/MTG-Arena-Tool/releases"
          rel="noopener noreferrer"
          target="_blank"
        >
          Releases
        </a>
      </div>
      <div
        className="footerColumn"
      >
        <div
          className="footerColumnTitle"
        >
          Support us
        </div>
        <a
          href="https://www.paypal.me/ManuelEtchegaray/10"
          rel="noopener noreferrer"
          target="_blank"
        >
          Donate
        </a>
        <a
          href="https://www.patreon.com/mtgatool"
          rel="noopener noreferrer"
          target="_blank"
        >
          Patreon
        </a>
      </div>
      <div
        className="footerColumn"
      >
        <div
          className="footerColumnTitle"
        >
          Developers
        </div>
        <a
          href="https://github.com/Manuel-777/MTG-Arena-Tool"
          rel="noopener noreferrer"
          target="_blank"
        >
          App Github
        </a>
        <a
          href="https://github.com/mtgatool/mtgatool-web/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Website Github
        </a>
      </div>
      <div
        className="footerColumn"
      >
        <div
          className="footerColumnTitle"
        >
          Places
        </div>
        <a
          href="https://discord.gg/K9bPkJy"
          rel="noopener noreferrer"
          target="_blank"
        >
          Discord
        </a>
        <a
          href="https://twitter.com/MEtchegaray7"
          rel="noopener noreferrer"
          target="_blank"
        >
          Twitter
        </a>
        <a
          href="https://magic.wizards.com/mtgarena"
          rel="noopener noreferrer"
          target="_blank"
        >
          MTG: Arena
        </a>
      </div>
    </div>
    <div
      className="footerWrapper"
    >
      <div
        className="footerText"
      >
        Portions of MTG Arena Tool are unofficial Fan Content permitted under the Fan Content Policy.
        <br />
        The literal and graphical information presented on this site about Magic: The Gathering, including card images, the mana symbols, and Oracle text, is copyright Wizards of the Coast, LLC, a subsidiary of Hasbro, Inc. MTG Arena Tool is not produced by, endorsed by, supported by, or affiliated with Wizards of the Coast.
        <br />
        ©Wizards of the Coast LLC.
        <br />
        All other content ©2020 MTG Arena Tool.
      </div>
    </div>
  </div>,
]
`;
